### PR TITLE
Clone the repo faster in Buildkite

### PR DESCRIPTION
**Pull request by mokagio**
By default, Buildkite runs a `git clone` that fetches the entire repository history, for the main repo as well as the submodules (e.g. `git clone -v -- git@github.com:wordpress-mobile/gutenberg-mobile.git .`)

This approach is too slow for a huge repository like ours. A faster strategy is to run a shallow clone, `git clone --depth=1 --shallows-submodules`. Buildkite allows doing this via environment variables and the environment hook, but our CI configuration blocks that for security reason.

I'm not sure what is the safest and more robust way to move forward. In the meantime, I found [a plugin that allows setting those environment variables](https://github.com/peakon/git-shallow-clone-buildkite-plugin) and copied the approach into [a plugin under our org](https://github.com/automattic/git-shallow-clone-buildkite-plugin). The reason for having our own plugin instead of relying and contributing to a community one is that it seems a safer approach. Hat tip @crazytonyli for pointing this out.

Untitled5
_The screenshot shows a gain of more than 1 minute during checkout._

## Testing

Green CI is enough here.

## Alternatives

I guess something we could do is make the behavior implemented in the plugin the default in our Buildkite agents setup. Maybe this has never come up because the git checkout has never been a big problem in other repos.

In the meantime, this does the job in the context of this pipeline.

---

**PR submission checklist**:

- [x] I have considered adding unit tests where possible. – N.A.
- [x] I have considered if this change warrants user-facing release notes [more info](https://github.com/wordpress-mobile/gutenberg-mobile/blob/trunk/Release-notes.md) and have added them to [RELEASE-NOTES.txt](https://github.com/wordpress-mobile/gutenberg-mobile/blob/trunk/RELEASE-NOTES.txt) if necessary. – N.A.

### CODE REVIEW

**Review by crazytonyli — APPROVED** [review]

**Comment by SiobhyB** [comment]
@mokagio, 👋 , following the ping [here](https://github.com/wordpress-mobile/gutenberg-mobile/pull/6024#issuecomment-1664159900), I wanted to provide some further context. We noticed that the `Build iOS RN XCFramework & Publish to S3` tests began to fail after this PR was merged.

- [Avoid navigation container dismissing the keyboard](https://github.com/wordpress-mobile/gutenberg-mobile/pull/6018) [[test reference](https://buildkite.com/automattic/gutenberg-mobile/builds/6710#0189ba6a-163a-48d4-86ff-1c7fdb9e863d)]:

<details>
 <summary>Error log here ⤵️</summary>

```
$ cd /usr/local/var/buildkite-agent/plugins/github-com-automattic-git-shallow-clone-buildkite-plugin-trunk3830653496
--
  | $ git clone -v -- https://github.com/automattic/git-shallow-clone-buildkite-plugin .
  | xcrun: error: invalid active developer path (/Library/Developer/CommandLineTools), missing xcrun at: /Library/Developer/CommandLineTools/usr/bin/xcrun
  | $ git clone -v -- https://github.com/automattic/git-shallow-clone-buildkite-plugin .
  | xcrun: error: invalid active developer path (/Library/Developer/CommandLineTools), missing xcrun at: /Library/Developer/CommandLineTools/usr/bin/xcrun
  | $ git clone -v -- https://github.com/automattic/git-shallow-clone-buildkite-plugin .
  | xcrun: error: invalid active developer path (/Library/Developer/CommandLineTools), missing xcrun at: /Library/Developer/CommandLineTools/usr/bin/xcrun
  | $ cd /usr/local/var/buildkite-agent/builds
  | $ cd /usr/local/var/buildkite-agent/builds/MV-MKE-X64-014/automattic/gutenberg-mobile
  | 🚨 Error: Failed to checkout plugin git-shallow-clone: exit status 1
```
</details>

- [Fix iOS Focus loop for RichText components](https://github.com/WordPress/gutenberg/pull/53217) [[test reference](https://buildkite.com/automattic/gutenberg-mobile/builds/6714#0189ba6d-68d9-4944-a83b-4819713fba4b)]

<details>
 <summary>Error log here ⤵️</summary>

```
$ cd /usr/local/var/buildkite-agent/plugins/github-com-automattic-a8c-ci-toolkit-buildkite-plugin-2-18-2994260596
--
  | $ git clone -v -- https://github.com/automattic/a8c-ci-toolkit-buildkite-plugin .
  | xcrun: error: invalid active developer path (/Library/Developer/CommandLineTools), missing xcrun at: /Library/Developer/CommandLineTools/usr/bin/xcrun
  | $ git clone -v -- https://github.com/automattic/a8c-ci-toolkit-buildkite-plugin .
  | xcrun: error: invalid active developer path (/Library/Developer/CommandLineTools), missing xcrun at: /Library/Developer/CommandLineTools/usr/bin/xcrun
  | $ git clone -v -- https://github.com/automattic/a8c-ci-toolkit-buildkite-plugin .
  | xcrun: error: invalid active developer path (/Library/Developer/CommandLineTools), missing xcrun at: /Library/Developer/CommandLineTools/usr/bin/xcrun
  | $ cd /usr/local/var/buildkite-agent/builds
  | $ cd /usr/local/var/buildkite-agent/builds/MV-MKE-X64-018/automattic/gutenberg-mobile
  | 🚨 Error: Failed to checkout plugin a8c-ci-toolkit: exit status 1
```
</details>

Following discussion at p1691052039405189/1691046441.873259-slack-C0126N6JH7Z, we decided to temporarily revert for the release branch (but not merge the revert to trunk) as we didn't want to block the release. We'd be grateful if you could help us look into why the tests began to fail so we can get a more permanent fix 🙇‍♀️

**Comment by AliSoftware** [comment]
@mokagio For the record, I would not recommend to use shallow clones, as from experience it often leads to edge cases in some git operations that become hard to debug. Which might explain the errors that Siobhan is seeing above.

---

What I would recommend instead to speed up git clones is to use our git cache mechanism that we use for our other repos, i.e.:
 - Have a [scheduled build set up via the pipeline's `.tf` file](https://github.com/Automattic/buildkite-ci/blob/trunk/src/buildkite/pipelines/wordpress-ios.tf#L63-L74) to run [`cache_repo`](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/blob/trunk/bin/cache_repo) via [a dedicated yaml](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/.buildkite/cache-builder.yml) on a daily basis, so that the git cache is reconstructed and uploaded regularly in S3.
 - Then [use our `git-s3-cache` plugin](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/.buildkite/pipeline.yml#L6-L8) on the steps of the repo's `pipeline.yml` so that Buildkite uses that git cache [as a `--reference`](https://github.com/Automattic/git-s3-cache-buildkite-plugin/blob/trunk/hooks/pre-checkout#L160) and starting point when doing the cloning, instead of having to clone the whole repo from scratch.

**Comment by mokagio** [comment]
🤦‍♂️ the `git-s3-cache` setup was so good and invisible that, despite having dealt to my fair share of shallow checkout issues in the CircleCI days, I totally forgot about it.

Thank you for the suggestion @AliSoftware. See

- https://github.com/wordpress-mobile/gutenberg-mobile/pull/6048
- https://github.com/Automattic/buildkite-ci/pull/246

**Pull request merged**
